Test engine implementation and module loader

Danesprite opened the pull request:
Re: #36.

Danesprite commented:
This is pretty much finished now. I still need to add some documentation and the script functionality I mentioned in the issue. I intend to use this to greatly simplify the Pocket Sphinx backend too.

I decided on a simpler way of handling dictation: any uppercase words are treated as dictation words. They will be received as lowercase by rules for consistency.

I was thinking that there could be a few options for the script. Maybe there could be an option for reading from standard input into `mimic()` somehow?

@LexiconCode I've actually tested this with Caster and it seems to work pretty well. I had to make a few modifications to *_caster.py* to get it working though.

LexiconCode commented:
Fantastic you can shoot either modified __caster.py_ or open up a pull request. I'll take some time testing as this should replace our current module verification system in caster.

LexiconCode commented:
My notes from experimenting
1. No other dragonfly/caster commands are available for dictation due to changing the engine to `engine("text")`. - Expected
2. Dragons free dictation and commands through still works as expected. 
3. With an unmodified caster.py everything works as expected when compared to the dragonfly master branch.
4. `engine.mimic` works as expected
```
engine.mimic("arch brov char ace format TESTING HELLO TESTING ace "
                 "word number one ace numb ten")
```
5. Caster did _not_ start WSR during my testing.

LexiconCode commented:
Can you define what you mean by standard input?

Danesprite commented:
Thanks for testing this! :+1:

I was testing by running *_caster.py* directly, not from Dragon, so that would be why WSR didn't start for you. Thanks for checking that it works alongside the other engines, I hadn't thought to check that.


> Can you define what you mean by standard input?

I was thinking of allowing something like the following in Git Bash/Cygwin:

```shell
echo "format TESTING\n again do" | python -m dragonfly --test _caster.py
```

Where `"format TESTING"` and `"again do"` would be mimicked with separate calls because of the new line. Without piping, the script would accept commands separated by new lines until interrupted with ctrl+c or the end of file character is received:

```shell
python -m dragonfly --test _caster.py
format TESTING
again do
```

LexiconCode commented:
`[open] command palette` throws an error vs `open command palette`. The same could be said for ` go to line one` function as expected but `go to line 1`.

I'm not all that surprised as is utilizes mimic. However how we approach testing using raw spec `go to line 1`. vs `go to line one` is an interesting question. Perhaps I'm overthinking it.

Danesprite commented:
Yep, mimic wouldn't accept specs like `"[open] command palette"`. As for `"go to line 1"`, I noticed that too. The error happens because dragonfly's `Integer` classes recognise word numbers, not digits.

Testing based on specs alone can be done, but you'd need to expand references and optionals (e.g. `<n>` and `[open]`) to get a list possible recognitions for it to work through `mimic()`. I don't think it's really worth the effort though. I suppose you could just remove any brackets with:
```Python
spec = spec.replace("[", "").replace("]", "")
```

I'll put some docs together for this and add something using my idea above. I think the engine tests should be added to the Travis CI build too.

Danesprite commented:
That is implemented now and I've added the tests to the CI build. I'll do the doc page and release version 0.10.0 tomorrow.

It's pretty much what I had in mind, but I've left it open to adding other commands later, perhaps for easily using RPC. I've tested it on Linux and Windows with both major versions of Python.

Here are some examples:
```Shell
# Help on available and specific commands
python -m dragonfly -h
python -m dragonfly test -h

# Simple test with caster using a pipe.
echo "arch brov char" | python -m dragonfly test  _caster.py

# Same test without the pipe.
python -m dragonfly test _caster.py
arch brov char

# Same test with quieter output.
echo "arch brov char" | python -m dragonfly test -q  _caster.py

# Test that the sphinx engine loads two modules and exit without checking input.
# Without -e/--engine, it will default to the "text" engine.
python -m dragonfly test -e sphinx --no-input module1.py module2.py
```

You can also set the logging level with `-l`/`--log-level`.

tylercal commented:
This pull request seems to change the behavior of how `dfly-loader-wsr.py ` loads modules. The Sphinx loader had condition that modules start with `_`
```python
if not (os.path.basename(path).startswith("_") and
                    os.path.splitext(path)[1] == ".py"):
```
while the WSR loader only verified that they were python files
```python
if not os.path.splitext(path)[1] == ".py":
                continue
```
But the combined behavior is that modules must start with an underscore? This breaks the examples in the examples directory.

Danesprite commented:
Thanks. I changed this to be in-line with which files are loaded by *natlink*. I can change the module loaders back to using the old behaviour. I think some of the newer example files will also need to be changed.

tylercal commented:
It would probably be nice for all the loaders to be consistent (unless someone is trying to use 2 in parallel).

Certainly cleaner to have the shared logic where you have it. I think the examples (and maybe some documentation) just need to be updated 😊

I’d be happy to open a PR for it, but it will take a while as I’m still working on getting my setup switched over from the original Drangonfly.

Danesprite commented:
So we'll stick with "_*.py", rename the examples and update the documentation? That sounds reasonable to me. It would also make the example files consistent with *natlink*. I like being able to disable cmd modules by just removing the underscore in the file name.

I think using two module loaders at once is unlikely, but I guess I could see use cases for that. It would be nice to have a `dragonfly load-modules <file>+ ...` command similar to the `dragonfly test <file>+` command added in this PR. It would be more flexible than the module loader scripts.

Let me know if you have issues migrating to *dragonfly2*. I'd like it to be as painless as possible really.

Danesprite commented:
@tylercal I've renamed the examples and adjusted the documentation for the module loaders (commits: 342c52b2 dbe4e7b2a).

I should probably  have made another issue for this, but I have mentioned the changes in the changelog.
